Design review: "I want a copy of folder X in every function"

Some users have asked for a common folder "X" to be copied into every one of their functions at build time.

i.e.

```
functions/
 + /fn1/
 + /fn2/
 + /fn3/
 + /common_folder/
```

In the scenario above - the user wants "common_folder" to be copied into or to be accessible within fn1, fn2 and fn3 and also expects the CLI to do that.

> My take - this is something which is not possible with Docker due to files outside the build context being inaccessible. It doesn't feel like the responsibility of the faas-cli templating system to get involved in additional user files.

Possible workaround: have user make use of a Makefile or batch file to copy functions into a temporary folder and build out the structure the way they want.

@stefanprodan suggested using "grunt" or a JavaScript task runner just like when you build a react website from a "src" folder into a "dest" folder.

@alexellis here are some alternative methods using make and bash

# Alternatives: 

## Makefile

```Makefile
.PHONY: build
build: 
    faas-cli build --shrinkwrap
    # maybe i can try to parse the stack file for the name of the functions and iterate here?
    cp -R ./common ./build/<function_name>/
    docker build  -t <prefix>/<function_name> ./build/<function_name>
```

Biggest downsides:
1. we need to hard code the copy commands to each function _or_ try to parse a yaml file in make
2. passing variables to makefiles is not easy, it can be done, but is very ugly
3. it is difficult to expose the various other flags in `faas-cli`
4. The author needs to re-implement the tagging and cannot easily reuse the information from the stack.yml


```Makefile
.FUNCTION_FILTER=""

.PHONY: build
build: 
    faas-cli build --shrinkwrap --filter="${.FUNCTION_FILTER}
    # maybe i can try to parse the stack file for the name of the functions, then apply the filter, and iterate?
    cp -R ./common ./build/<function_name>
    docker build  -t <prefix>/<function_name> ./build/<function_name>
```

usage would look like 

```sh
➜  echo-fn git:(master) ✗ make build
faas-cli build --shrinkwrap --filter=""""
[0] > Building go-echo.
Clearing temporary build folder: ./build/go-echo/
Preparing ./go-echo/ ./build/go-echo//function
Building: alexellis2/go-echo:latest with golang-http template. Please wait..
go-echo shrink-wrapped to ./build/go-echo/
[0] < Building go-echo done.
[0] worker done.
cp -R ./common ./build/go-echo

➜  echo-fn git:(master) ✗ make build .FUNCTION_FILTER=foo
faas-cli build --shrinkwrap --filter="foo"
No functions matching --filter/--regex were found in the YAML file
make: *** [build] Error 1
```


## Bash

```bash
#!  /usr/bin/env bash

faas-cli build --shrinkwrap $@
# maybe i can try to parse the stack file for the name of the functions and iterate?
cp -R ./common ./build/<function_name>
docker build  -t <prefix>/<function_name> ./build/<function_name>
```

This script is simpler than the `Makefile` because we can blanket pass extra args to `faas-cli` via the `$@`. However, it still has the hard coded function name

Biggest Downsides:
1. we need to hard code the copy commands to each function _or_ try to parse a yaml file in make
2. The author needs to re-implement the tagging and cannot easily reuse the information from the stack.yml

## Custom template

One more option that I came up with is create custom template with the common code.  The flow would be to 

1. pull the desired template
2. create/copy the code into the template
3. commit the template folder with the repo

Biggest downsides:
1. discoverability and maintenance of the common code is now harder, I would not expect project code to be in the template dir, so this would have to be documented in the project
2. related to discoverability, there is almost not chance an editor/IDE is going to be able to index the code in the common folder when it is in the template
3. if the project has multiple functions that want to use different templates, this would result in copying the code multiple times
4. The author needs to re-implement the tagging and cannot easily reuse the information from the stack.yml

## Bash Alternative

After thinking about the "Custom Template" approach, I came up with this alternative bash approach that copies the common folder into the template prior to being built.  
This script is also more dynamic that the others

```sh
#!  /usr/bin/env bash

for d in ./template/*; 
do 
    cp -R ./common "$d"
done

faas-cli build --shrinkwrap $@

for d in ./build/*;
do
    echo "Building ${d:8}"
    docker build -t <prefix>/${d:8}:latest "$d"
done
```



Biggest downsides:
1. This really only makes sense for files/folders that are shared among all functions of a project, there is not a lot of control in the sense of "functions A and B need to share `common` while C and D need to share `models`", for example. 
2. The author needs to re-implement the tagging and cannot easily reuse the information from the stack.yml
3. requires the project author to be comfortable with bash loops and string manipulation


--- 
I am fairly comfortable with Bash, so, I would probably advocate for the final option, if someone was going to pursue this direction.  It is the most flexible for being directly copied into a new project

--- a/builder/build.go
+++ b/builder/build.go
@@ -23,7 +23,7 @@ import (
 const AdditionalPackageBuildArg = "ADDITIONAL_PACKAGE"
 
 // BuildImage construct Docker image from function parameters
-func BuildImage(image string, handler string, functionName string, language string, nocache bool, squash bool, shrinkwrap bool, buildArgMap map[string]string, buildOptions []string, tagMode schema.BuildFormat, buildLabelMap map[string]string, quietBuild bool) error {
+func BuildImage(image string, handler string, functionName string, language string, nocache bool, squash bool, shrinkwrap bool, buildArgMap map[string]string, buildOptions []string, tagMode schema.BuildFormat, buildLabelMap map[string]string, quietBuild bool, copyExtraPaths []string) error {
 
 	if stack.IsValidTemplate(language) {
 		pathToTemplateYAML := fmt.Sprintf("./template/%s/template.yml", language)
@@ -47,7 +47,7 @@ func BuildImage(image string, handler string, functionName string, language stri
 			return fmt.Errorf("building %s, %s is an invalid path", imageName, handler)
 		}
 
-		tempPath, buildErr := createBuildContext(functionName, handler, language, isLanguageTemplate(language), langTemplate.HandlerFolder)
+		tempPath, buildErr := createBuildContext(functionName, handler, language, isLanguageTemplate(language), langTemplate.HandlerFolder, copyExtraPaths)
 		fmt.Printf("Building: %s with %s template. Please wait..\n", imageName, language)
 		if buildErr != nil {
 			return buildErr
@@ -164,7 +164,7 @@ type dockerBuild struct {
 const defaultHandlerFolder = "function"
 
 // createBuildContext creates temporary build folder to perform a Docker build with language template
-func createBuildContext(functionName string, handler string, language string, useFunction bool, handlerFolder string) (string, error) {
+func createBuildContext(functionName string, handler string, language string, useFunction bool, handlerFolder string, copyExtraPaths []string) (string, error) {
 	tempPath := fmt.Sprintf("./build/%s/", functionName)
 	fmt.Printf("Clearing temporary build folder: %s\n", tempPath)
 
@@ -225,9 +225,53 @@ func createBuildContext(functionName string, handler string, language string, us
 		}
 	}
 
+	for _, extraPath := range copyExtraPaths {
+		extraPathAbs, err := pathInScope(extraPath, ".")
+		if err != nil {
+			return tempPath, err
+		}
+		// Note that if useFunction is false, ie is a `dockerfile` template, then
+		// functionPath == tempPath, the docker build context, not the `function` handler folder
+		// inside the docker build context
+		copyErr := CopyFiles(
+			extraPathAbs,
+			filepath.Clean(path.Join(functionPath, extraPath)),
+		)
+
+		if copyErr != nil {
+			return tempPath, copyErr
+		}
+	}
+
 	return tempPath, nil
 }
 
+// pathInScope returns the absolute path to `path` and ensures that it is located within the
+// provided scope. An error will be returned, if the path is outside of the provided scope.
+func pathInScope(path string, scope string) (string, error) {
+	scope, err := filepath.Abs(filepath.FromSlash(scope))
+	if err != nil {
+		return "", err
+	}
+
+	abs, err := filepath.Abs(filepath.FromSlash(path))
+	if err != nil {
+		return "", err
+	}
+
+	if abs == scope {
+		return "", fmt.Errorf("forbidden path appears to equal the entire project: %s (%s)", path, abs)
+	}
+
+	if strings.HasPrefix(abs, scope) {
+		return abs, nil
+	}
+
+	// default return is an error
+	return "", fmt.Errorf("forbidden path appears to be outside of the build context: %s (%s)", path, abs)
+}
+
+// appears to be unused???
 func dockerBuildFolder(functionName string, handler string, language string) string {
 	tempPath := fmt.Sprintf("./build/%s/", functionName)
 	fmt.Printf("Clearing temporary build folder: %s\n", tempPath)

--- a/builder/build_test.go
+++ b/builder/build_test.go
@@ -2,6 +2,7 @@ package builder
 
 import (
 	"fmt"
+	"path/filepath"
 	"strings"
 	"testing"
 
@@ -467,6 +468,69 @@ func Test_deDuplicate(t *testing.T) {
 
 				if found == false {
 					t.Errorf("Slices differ in values  - wanted: %s, found %s", strings.Join(test.expectedStrings, " "), strings.Join(uniqueStrings, " "))
+				}
+			}
+		})
+	}
+}
+
+func Test_pathInScope(t *testing.T) {
+	root, err := filepath.Abs(".")
+	if err != nil {
+		t.Fatalf("unexpected error during test setup: %s", err)
+	}
+
+	cases := []struct {
+		name         string
+		path         string
+		expectedPath string
+		err          bool
+	}{
+		{
+			name:         "can copy folders without any relative path prefix",
+			path:         "common/models/prebuilt",
+			expectedPath: filepath.Join(root, "common/models/prebuilt"),
+		},
+		{
+			name:         "can copy folders with relative path prefix",
+			path:         "./common/data/cleaned",
+			expectedPath: filepath.Join(root, "./common/data/cleaned"),
+		},
+		{
+			name: "error if path equals the current directory",
+			path: "./",
+			err:  true,
+		},
+		{
+			name: "error if relative path moves out of the current directory",
+			path: "../private",
+			err:  true,
+		},
+		{
+			name: "error if absolute path is outside of the current directory",
+			path: "/private/common",
+			err:  true,
+		},
+		{
+			name: "error if relative path moves out of the current directory, even when hidden in the middle of the path",
+			path: "./common/../../private",
+			err:  true,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			abs, err := pathInScope(tc.path, root)
+			switch {
+			case tc.err && err == nil:
+				t.Fatalf("expected error but got none")
+			case tc.err && !strings.HasPrefix(err.Error(), "forbidden path"):
+				t.Fatalf("expected forbidden path error, got \"%s\"", err)
+			case !tc.err && err != nil:
+				t.Fatalf("unexpected err \"%s\"", err)
+			default:
+				if abs != tc.expectedPath {
+					t.Fatalf("expected path %s, got %s", tc.expectedPath, abs)
 				}
 			}
 		})

--- a/commands/build.go
+++ b/commands/build.go
@@ -28,6 +28,7 @@ var (
 	buildArgs        []string
 	buildArgMap      map[string]string
 	buildOptions     []string
+	copyExtra        []string
 	tagFormat        schema.BuildFormat
 	buildLabels      []string
 	buildLabelMap    map[string]string
@@ -52,6 +53,7 @@ func init() {
 	buildCmd.Flags().StringArrayVarP(&buildOptions, "build-option", "o", []string{}, "Set a build option, e.g. dev")
 	buildCmd.Flags().Var(&tagFormat, "tag", "Override latest tag on function Docker image, accepts 'latest', 'sha', 'branch', or 'describe'")
 	buildCmd.Flags().StringArrayVar(&buildLabels, "build-label", []string{}, "Add a label for Docker image (LABEL=VALUE)")
+	buildCmd.Flags().StringArrayVar(&copyExtra, "copy-extra", []string{}, "Extra paths that will be copied into the function build context")
 	buildCmd.Flags().BoolVar(&envsubst, "envsubst", true, "Substitute environment variables in stack.yml file")
 	buildCmd.Flags().BoolVar(&quietBuild, "quiet", false, "Perform a quiet build, without showing output from Docker")
 	buildCmd.Flags().BoolVar(&disableStackPull, "disable-stack-pull", false, "Disables the template configuration in the stack.yml")
@@ -75,6 +77,7 @@ var buildCmd = &cobra.Command{
 				 [--parallel PARALLEL_DEPTH]
 				 [--build-arg KEY=VALUE]
 				 [--build-option VALUE]
+				 [--copy-extra PATH]
 				 [--tag <sha|branch|describe>]`,
 	Short: "Builds OpenFaaS function containers",
 	Long: `Builds OpenFaaS function containers either via the supplied YAML config using
@@ -185,8 +188,9 @@ func runBuild(cmd *cobra.Command, args []string) error {
 			buildOptions,
 			tagFormat,
 			buildLabelMap,
-			quietBuild)
-
+			quietBuild,
+			copyExtra,
+		)
 		if err != nil {
 			return err
 		}
@@ -230,8 +234,8 @@ func build(services *stack.Services, queueDepth int, shrinkwrap, quietBuild bool
 				if len(function.Language) == 0 {
 					fmt.Println("Please provide a valid language for your function.")
 				} else {
-
 					combinedBuildOptions := combineBuildOpts(function.BuildOptions, buildOptions)
+					combinedExtraPaths := mergeSlice(services.StackConfiguration.CopyExtraPaths, copyExtra)
 					err := builder.BuildImage(function.Image,
 						function.Handler,
 						function.Name,
@@ -243,7 +247,9 @@ func build(services *stack.Services, queueDepth int, shrinkwrap, quietBuild bool
 						combinedBuildOptions,
 						tagFormat,
 						buildLabelMap,
-						quietBuild)
+						quietBuild,
+						combinedExtraPaths,
+					)
 
 					if err != nil {
 						errors = append(errors, err)

--- a/stack/schema.go
+++ b/stack/schema.go
@@ -70,6 +70,13 @@ type Configuration struct {
 // StackConfiguration for the overall stack.yml
 type StackConfiguration struct {
 	TemplateConfigs []TemplateSource `yaml:"templates"`
+	// CopyExtraPaths specifies additional paths (relative to the stack file) that will be copied
+	// into the functions build context, e.g. specifying `"common"` will look for and copy the
+	// "common/" folder of file in the same root as the stack file.  All paths must be contained
+	// within the project root defined by the location of the stack file.
+	//
+	// The yaml uses the shorter name `copy` to make it easier for developers to read and use
+	CopyExtraPaths []string `yaml:"copy"`
 }
 
 // TemplateSource for build templates

